### PR TITLE
(WIP) feat/data-manager: add support for idempotent operations

### PR DESCRIPTION
**Pull request by madadam**

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @Fraser999

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/personas/data_manager/mod.rs [hunk 1]
```diff
@@ -392,8 +392,16 @@ impl DataManager {
...
-        let res = if self.chunk_store.has(&data_id) {
-            Err(ClientError::DataExists)
+        let res = if let Ok(old_data) = self.chunk_store.get(&data_id) {
+            if old_data == data {
+                // Putting the same data more than once is allowed to support idempotence.
+                routing_node
+                    .send_put_mdata_response(dst, src, Ok(()), msg_id)?;
+                return Ok(());
+            } else {
+                // But putting different data with the same id is not.
```
Comment (maqi) [inline src/personas/data_manager/mod.rs]: what if the client sends seq of request with: put(v_0), update(v_1), put(v_0)  ?

File: tests/data_manager.rs [hunk 2]
```diff
@@ -1156,6 +1157,87 @@ fn mutable_data_operations_with_churn() {
...
+
+    client.ensure_connected(&mut nodes);
+    client.create_account(&mut nodes);
+
+    // Sequentially putting the same mdata more than once.
+    let data = test_utils::gen_mutable_data(10000, 2, *client.signing_public_key(), &mut rng);
+    unwrap!(client.put_mdata_response(data.clone(), &mut nodes));
+    unwrap!(client.put_mdata_response(data, &mut nodes));
+
+    // Concurrently putting the same mdata more than once.
```
Comment (maqi) [inline tests/data_manager.rs]: better using two clients to carry out this concurrently put ?

File: tests/data_manager.rs [hunk 3]
```diff
@@ -1156,6 +1157,87 @@ fn mutable_data_operations_with_churn() {
...
+
+    // Sequentially mutating the data with the same actions more than once.
+    let actions = test_utils::gen_mutable_data_entry_actions(&data, 2, &mut rng);
+    unwrap!(client.mutate_mdata_entries_response(*data.name(),
+                                                 data.tag(),
+                                                 actions.clone(),
+                                                 &mut nodes));
+    unwrap!(client.mutate_mdata_entries_response(*data.name(), data.tag(), actions, &mut nodes));
+
+    // Concurrently mutating the data with the same actions more than once.
```
Comment (maqi) [inline tests/data_manager.rs]: using two clients to carry out this concurrently mutate can be more convincing?

File: tests/data_manager.rs [hunk 4]
```diff
@@ -1156,6 +1157,87 @@ fn mutable_data_operations_with_churn() {
...
+                }
+                if msg_id == msg_id_1 {
+                    responded_1 = true
+                }
+            }
+        }
+    }
+
+    assert!(responded_0);
+    assert!(responded_1);
```
Comment (maqi) [inline tests/data_manager.rs]: why not just put msg_ids in BTreeSet, then only need to insert received msg_id assert tree_set are identical?
same comment to the above.

**Comment by ustulation** [comment]
We are going for a different paradigm that does not affect routing and vaults: MAID-2146 in JIRA.

**Pull request closed without merging**